pytket/extensions/cutensornet/general_state/tensor_network_state.py

```diff
@@ -0,0 +1,466 @@
+
+
+class GeneralState:
+    """Handles cuTensorNet tensor network state object."""
+
+    def __init__(
+        self,
+        circuit: Circuit,
+        libhandle: CuTensorNetHandle,
+        loglevel: int = logging.INFO,
+    ) -> None:
+        """Constructs a tensor network state representation from a pytket circuit.
+
+        Note:
+            Circuit should not contain boxes - only explicit gates with specific unitary
+            matrix representation available in pytket.
+
+        Note:
+            If present, implicit wire swaps are replaced with explicit SWAP gates.
+
+        Args:
+            circuit: A pytket circuit to be converted to a tensor network.
+            libhandle: cuTensorNet handle.
+            loglevel: Internal logger output level.
+        """
+        self._logger = set_logger("GeneralState", loglevel)
+        self._circuit = circuit
+        self._circuit.replace_implicit_wire_swaps()
+        self._handle = libhandle.handle
+        self._dev = libhandle.dev
+
+        libhandle.print_device_properties(self._logger)
+
+        num_qubits = circuit.n_qubits
+        dim = 2  # We are always dealing with qubits, not qudits
+        qubits_dims = (dim,) * num_qubits  # qubit size
+        self._logger.debug(f"Converting a quantum circuit with {num_qubits} qubits.")
+        data_type = cq.cudaDataType.CUDA_C_64F  # for now let that be hard-coded
+
+        # These are only required when doing preparation and evaluation.
+        self._stream = None
+        self._scratch_space = None
+        self._work_desc = None
+
+        self._state = cutn.create_state(
+            self._handle, cutn.StatePurity.PURE, num_qubits, qubits_dims, data_type
+        )
+        self._mutable_gates_map = {}
+        self._gate_tensors = []
+        for com in circuit.get_commands():
+            gate_unitary = com.op.get_unitary().astype("complex128", copy=False)
+            # Transpose is needed because of the way cuTN stores tensors.
+            # See https://github.com/NVIDIA/cuQuantum/discussions/124
+            # #discussioncomment-8683146 for details.
+            self._gate_tensors.append(
+                cp.asarray(gate_unitary)
+                .T.astype(dtype="complex128", order="F")
+                .reshape([2] * (2 * com.op.n_qubits), order="F")
+            )
+            gate_strides = 0  # Always 0?
+            gate_qubit_indices = tuple(
+                self._circuit.qubits.index(qb) for qb in com.qubits
+            )
+            gate_id = cutn.state_apply_tensor(
+                self._handle,
+                self._state,
+                com.op.n_qubits,
+                gate_qubit_indices,
+                self._gate_tensors[-1].data.ptr,
+                gate_strides,
+                1,
+                0,
+                1,
+            )
+            if com.opgroup is not None:
+                self._mutable_gates_map[com.opgroup] = gate_id
```

Comment:
In fact, currently it can't be working, since `immutable=1` in the call to `state_apply_tensor_operator`. I'm comenting out anything that has to do with `self._mutable_gates_map`, and we can recover it at some later point if necessary.